src/cpp/session/modules/data/DataViewer.cpp

```diff
@@ -327,6 +327,28 @@ SEXP findInNamedEnvir(const std::string& envir, const std::string& name)
    return obj == R_UnboundValue ? nullptr : obj;
 }
 
+SEXP getNamedEnvir(const std::string& envir)
+{
+   SEXP env = nullptr;
+   r::sexp::Protect protect;
+
+   // shortcut for unbound environment
+   if (envir == kNoBoundEnv)
+      return nullptr;
+
+   // use the global environment or resolve environment name
+   if (envir.empty() || envir == "R_GlobalEnv")
+      return R_GlobalEnv;
+   
+   r::exec::RFunction(".rs.safeAsEnvironment", envir).call(&env, &protect);
```

Comment:
Since this uses `as.environment()`, this only finds packages which are currently attached (doesn't work for loaded packages; needs to use the search path name e.g. 'package:stats'). Is that sufficient? Can we document that restriction as a comment?

Reply:
Perhaps `CachedFrame` should also keep the actual environment in addition (instead of) its name, like the object explorer code does: 

```
> e <- new.env()
> e$x <- list(y = 3)
> with(e, View(x))
> e$x <- list(y = 4)
> ls.str(.rs.explorer.cache)
b1d85181a3494e4b85278bb86e220673 : List of 5
 $ object  :List of 1
 $ name    : chr "x"
 $ title   : chr "x"
 $ language: chr "R"
 $ envir   :<environment: 0x123fa6e10> 
> e
<environment: 0x123fa6e10>
```

with the View updated to 4: 

<img width="560" alt="image" src="https://user-images.githubusercontent.com/2625526/189337302-278dff4d-a42a-4658-a821-2731871bb2cb.png">



whereas this does not happen to the data viewer: 

```
> e <- new.env()
> e$x <- data.frame(y = 5)
> with(e, View(x))
> e$x <- data.frame(y = 6)
> ls.str(.rs.CachedDataEnv)
254C01D8 : 'data.frame':	1 obs. of  1 variable:
 $ y: num 5
```

and the view stays on 5: 

<img width="373" alt="image" src="https://user-images.githubusercontent.com/2625526/189339061-e2a30985-e5bb-41e3-bd99-f288a611413d.png">

Reply:
Now, I guess keeping the environment around, as in `.rs.explorer.cache` prevents its garbage collection. Maybe we can keep a weak reference of it so that: 

 - if it's kept around and modified (as in the examples above), then we can update the `View`
 - otherwise the object is final, similar to when we set the environment to "empty env" when it comes from evaluating some expression

Reply:
I never really got weak references, but the documentation in `rlang` does a good job of explaining them: https://rlang.r-lib.org/reference/new_weakref.html

Reply:
One other consideration: are these objects being serialized at all? If so, will weakrefs survive serialization? (I think we might be saving + loading some data for cached dataframes at least)

Reply:
The `CachedFrame` objects are not serialized, only the `.rs.CachedDataEnv` IIUC. 

```cpp
/*
 * Data Viewer caching overview
 * ----------------------------
 *
 * For each object being viewed, there are three copies to consider:
 * 
 * ORIGINAL:
 *    The original object on which the user invoked View(). This object may or
 *    may not exist; for example, View(cars) binds a viewer to the 'cars' 
 *    object in 'package:datasets', but View(rbind(cars,cars)) binds a viewer
 *    to a temporary object that doesn't exist anywhere but in the viewer.
 *
 *    When the original object exists, and no sorting or filtering is applied,
 *    requests for data are met by pulling data from the original object. We
 *    also watch original objects; when they're replaced in their hosting
 *    environments (assuming those environments are named), a client event is
 *    emitted.
 *
 * CACHED:
 *    Because the original object may be temporary (and, even if not, can be
 *    deleted at any time), we always have a cached copy of the object 
 *    available. 
 *
 *    The environment .rs.CachedDataEnv contains the cached objects. These
 *    objects have randomly generated cache keys. 
 *
 *    When the session suspends/resumes, the contents of the cache environment
 *    are written as individual .RData files to the user scratch folder. This
 *    allows us to reload the data for viewing afterwards.
 *
 *    The client is responsible for letting the server know when the viewer has
 *    closed; when this happens, the server removes the in-memory and disk 
 *    cache entries.
 *
 * WORKING:
 *    As the user orders, filters, and searches data, it's typical to follow a
 *    narrowing approach--e.g. first show only "Housewares", then only
 *    housewares between $10-$25, then only housewares between $10-25 and
 *    matching the text "eggs".
 *    
 *    In order to avoid re-ordering and re-filtering the entire dataset every
 *    time a new set of rows is requested, we keep a "working copy" of the
 *    object in a second environment, .rs.WorkingDataEnv, using the same cache
 *    keys.
 *    
 *    When a request for data arrives, we check to see if the data requested is
 *    a subset of the data already in our working copy. If it is, we use the
 *    working copy as a starting position rather than the original or cached
 *    object.
 *
 *    This allows us to efficiently perform operations on very large datasets
 *    once they've been winnowed down to smaller objects using searches and
 *    filters.
 */    
```

So I guess what I'm proposing here is to lift the `(assuming those environments are named)` restriction in the `ORIGINAL` section.

Reply:
Maybe this is a follow up issue/pr because it may not be trivial, `getGridData()` uses environment and object names in addition to `cache_key`.

Reply:
Ahh gotcha. Okay that makes sense! Seems fine to leave it for a separate issue.